Tests/Simulation/SimulatedLoadTests.cs

```diff
@@ -119,10 +119,121 @@ async Task ActAsync(SimulatedDevice device, TimeSpan startOffset)
             }
         }
 
+        // Scenario:
+        // - ABP devices
+        // - OTAA devices
+        // - Sending confirmed/unconfirmed messages
+        // - Goal: N devices in parallel based on configuration
+        [Fact]
+        public async Task Multiple_ABP_and_OTAA_Simulated_Devices_Confirmed()
+        {
+            // amount of messages to send per device (without warm-up phase)
+            const int scenarioMessagesPerDevice = 10;
+
+            // amount of devices to send data in parallel
+            const int scenarioDeviceStepSize = 15;
+
+            // amount of devices to send data in parallel for the warm-up phase
+            const int warmUpDeviceStepSize = 2;
+
+            // amount of messages to send before device join is to occur
+            const int messagesBeforeJoin = 5;
+
+            // amount of Unconfirmed messges to send before Confirmed message is to occur
+            const int messagesBeforeConfirmed = 5;
+
+            // amount of seconds to wait between sends in warmup phase
+            var delayWarmup = TimeSpan.FromSeconds(5);
+
+            var simulatedAbpDevices = InitializeSimulatedDevices(TestFixtureSim.DeviceRange2000_ABP_FullLoad);
+            var simulatedOtaaDevices = InitializeSimulatedDevices(TestFixtureSim.DeviceRange3000_OTAA_FullLoad);
+            Assert.Equal(simulatedAbpDevices.Count, simulatedOtaaDevices.Count);
+            Assert.True(simulatedOtaaDevices.Count < 50, "Simulator does not work for more than 50 of each devices (due to IoT Edge connection mode). To go beyond 100 device clients, use edge hub environment variable 'MaxConnectedClients'.");
+            Assert.True(messagesBeforeConfirmed <= messagesBeforeJoin, "OTAA devices should send all messages as confirmed messages.");
+
+            // 1. picking devices send an initial message (warm device cache in LNS module)
+            foreach (var devices in simulatedAbpDevices.Batch(warmUpDeviceStepSize))
+            {
+                await Task.WhenAll(from device in devices select SendWarmupMessageAsync(device));
+                await Task.Delay(delayWarmup);
+            }
+
+            async Task SendWarmupMessageAsync(SimulatedDevice device)
+            {
+                await Task.Delay(TimeSpan.FromMilliseconds(RandomNumberGenerator.GetInt32(10, 250)));
+                this.logger.LogInformation("[WARM-UP] {DeviceId}", device.LoRaDevice.DeviceID);
+                // warm-up messages should not contain the message fragment (we do not want to include them in the message count)
+                using var request = WaitableLoRaRequest.CreateWaitableRequest(device.CreateConfirmedDataUpMessage("foo"));
+                await device.SendDataMessageAsync(request);
+            }
+
+            // 2. ABP and OTAA devices send messages
+            const int initialMessageId = 1;
+            for (var messageId = initialMessageId; messageId < initialMessageId + scenarioMessagesPerDevice; ++messageId)
```

Comment:
[nit] Is there a technical reason why you try to send messages for both ABP and OTAA devices in a single for-loop? I find that it complicates the parameters we need to validate (e.g. messagesBeforeConfirmed <= messagesBeforeJoin) and the logic (e.g. for OTAA if messageId < messagesBeforeJoin + initialMessageId we don't send anything)

Reply:
In any case I would need to duplicate a loop (if I send ABP and OTAA separately it would need to be the loop over message IDs). Now, to make sure that I send OTAA and ABP for the same message ID at the same time, it needs to be in the same loop. We could argue whether it would be good enough to send OTAA and ABP after each other, but I decided to just draw a line somewhere and leave the old test as it was (with respect to its "flow").

No strong opinion here though, we can discuss whether the message ID is still a useful concept for that test at all.

Reply:
As per our discussion, I get it now. Originally, I had it as you suggested, but that did not work as expected due to some details around materialization of the Enumerables

Reply:
I see, in that case I think we can resolve, it was more of a nit comment in any case.